File: Data-Structures-Algorithms/Trees/Binary_Trees/symmetric-tree.cpp
/*
https://leetcode.com/problems/symmetric-tree/
*/

# include <bits/stdc++.h>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {

public:
    bool isMirror(TreeNode* tl, TreeNode* tr) {
        
        if (tl == NULL && tr == NULL)
            return true;
        
        if (tl == NULL || tr == NULL)
            return false;
        
        return (tl -> val == tr -> val) 
        && isMirror(tl -> left, tr -> right) && isMirror(tl -> right, tr -> left) ;
        
    }

    bool isSymmetric_recursive(TreeNode* root) {
        
        if (root == NULL)
            return true;
        
        return isMirror(root -> left, root -> right);
        
    }

    bool isSymmetric_iterative(TreeNode* root) {
        
        if (root == NULL)
            return true;
        
        queue<TreeNode*> ql;
        queue<TreeNode*> qr;
        
        TreeNode *l, *r;
        
        ql.push(root -> left);
        qr.push(root -> right);
        
        while ( !ql.empty() ) {
            
            l = ql.front();     ql.pop();           
            r = qr.front();     qr.pop();
            
            if (l == NULL && r == NULL)
                continue;
            
            if (l == NULL || r == NULL)
                return false;
            
            if (l -> val != r -> val)
                return false;
            
            ql.push(l -> left);     qr.push(r -> right);
            ql.push(l -> right);    qr.push(r -> left);
        }
            
        return true;        
    }
};

int main() {

    /*
    6
    1 2 2 3 4 4 3
    6
    1 2 2 null 3 null 3
    */

    /*
    1
    0
    */


    return 0;
}

File: Data-Structures-Algorithms/Trees/Binary-Trees/symmetric-tree.java
// https://leetcode.com/problems/symmetric-tree/description/?envType=study-plan-v2&envId=top-interview-150

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {

    public boolean isSymmetric(TreeNode root) {

        if (root == null) {
            return true;
        }

        Queue<TreeNode> ql = new LinkedList<TreeNode>();
        Queue<TreeNode> qr = new LinkedList<TreeNode>();

        ql.add(root.left);
        qr.add(root.right);

        TreeNode l, r;

        while (!ql.isEmpty()) {

            l = ql.poll();
            r = qr.poll();

            if (l == null && r == null) {
                continue;
            }
            if (l == null || r == null) {
                return false;
            }
            if (l.val != r.val) {
                return false;
            }

            ql.add(l.left);    qr.add(r.right);
            ql.add(l.right);   qr.add(r.left);
        }

        return true;
        
    }

}

class Solution2 {

    public boolean isSame(TreeNode p, TreeNode q) {
        if (p == null && q == null) {
            return true;
        }
        if (p == null || q == null) {
            return false;
        }
        return p.val == q.val && isSame(p.left, q.right) && isSame(p.right, q.left);
    }

    public boolean isSymmetric(TreeNode root) {
        return isSame(root.left, root.right);
    }
}